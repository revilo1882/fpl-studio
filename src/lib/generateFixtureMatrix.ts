import type { Fixtures, Team, BootstrapData } from '@/types/fpl'

import { calculateAttractivenessScore } from './attractivenessScore'
import { calculateDynamicFDR } from './fdr'

export type SingleFixture = {
	label: string
	difficulty: number
	opponentName: string
	kickoffTime: string | null
}

export type FixtureCell = SingleFixture[]

export type IGenerateFixtureMatrix = {
	teamNames: string[]
	fixtureMatrix: FixtureCell[][]
	scores: number[]
}
export type DifficultyType = 'fpl' | 'overall' | 'attack' | 'defence'

type GenerateFixtureMatrixProps = {
	teams: Team[]
	fixtures: Fixtures
	bootstrapData: BootstrapData
	firstGameweek: number
	numberOfGameweeks: number
	difficultyType: DifficultyType
}

export const generateFixtureMatrix = async ({
	teams,
	fixtures,
	firstGameweek,
	numberOfGameweeks,
	difficultyType,
}: GenerateFixtureMatrixProps): Promise<IGenerateFixtureMatrix> => {
	const teamMap = new Map(teams.map((team) => [team.id, team]))
	const teamNames = teams.map((team) => team.name)

	// For Studio views, pre-calculate FDR ratings
	const studioFDRCache: Map<string, number> = new Map()

	if (difficultyType !== 'fpl') {
		const fixturesForCalculation: Array<{
			homeTeam: Team
			awayTeam: Team
			gameweek: number
			fixtureId: number
		}> = []

		// Collect all fixtures that need calculation
		for (
			let gameweek = firstGameweek;
			gameweek < firstGameweek + numberOfGameweeks;
			gameweek++
		) {
			const gameweekFixtures = fixtures.filter((fixture) => fixture.event === gameweek)

			for (const fixture of gameweekFixtures) {
				const homeTeam = teamMap.get(fixture.team_h)
				const awayTeam = teamMap.get(fixture.team_a)

				if (homeTeam && awayTeam) {
					fixturesForCalculation.push({
						homeTeam,
						awayTeam,
						gameweek,
						fixtureId: fixture.id,
					})
				}
			}
		}

		// Calculate simplified FDR for all fixtures
		try {
			const fdrResults = await Promise.all(
				fixturesForCalculation.map(async ({ homeTeam, awayTeam, gameweek, fixtureId }) => {
					const [homeFDR, awayFDR] = await Promise.all([
						calculateDynamicFDR(homeTeam, awayTeam, fixtures, teams, true),
						calculateDynamicFDR(homeTeam, awayTeam, fixtures, teams, false),
					])

					// Extract the difficulty score based on type
					let homeDifficulty: number
					let awayDifficulty: number

					switch (difficultyType) {
						case 'attack':
							homeDifficulty = homeFDR.attacking
							awayDifficulty = awayFDR.attacking
							break
						case 'defence':
							homeDifficulty = homeFDR.defensive
							awayDifficulty = awayFDR.defensive
							break
						case 'overall':
						default:
							homeDifficulty = homeFDR.overall
							awayDifficulty = awayFDR.overall
							break
					}

					return {
						homeKey: `${homeTeam.id}-${fixtureId}-${gameweek}`,
						awayKey: `${awayTeam.id}-${fixtureId}-${gameweek}`,
						homeDifficulty,
						awayDifficulty,
					}
				}),
			)

			// Cache the results
			for (const { homeKey, awayKey, homeDifficulty, awayDifficulty } of fdrResults) {
				studioFDRCache.set(homeKey, homeDifficulty)
				studioFDRCache.set(awayKey, awayDifficulty)
			}
		} catch (error) {
			console.error('Error calculating simplified FDR, falling back to FPL ratings:', error)
		}
	}

	const fixtureMatrix: FixtureCell[][] = teams.map((team) => {
		const row: FixtureCell[] = []

		for (
			let gameweek = firstGameweek;
			gameweek < firstGameweek + numberOfGameweeks;
			gameweek++
		) {
			const teamFixtures = fixtures
				.filter(
					(fixture) =>
						fixture.event === gameweek &&
						(fixture.team_h === team.id || fixture.team_a === team.id),
				)
				.sort(
					(a, b) =>
						new Date(a.kickoff_time ?? '').getTime() -
						new Date(b.kickoff_time ?? '').getTime(),
				)

			if (teamFixtures.length === 0) {
				row.push([{ label: '-', difficulty: 0, opponentName: 'Blank', kickoffTime: null }])
				continue
			}

			const fixturesForWeek: FixtureCell = teamFixtures.map((fixture) => {
				const isHome = fixture.team_h === team.id
				const opponentId = isHome ? fixture.team_a : fixture.team_h
				const opponent = teamMap.get(opponentId)
				const opponentShort = opponent?.short_name ?? '?'
				const label = `${opponentShort} (${isHome ? 'H' : 'A'})`

				let difficulty: number

				if (!opponent) {
					difficulty = isHome ? fixture.team_h_difficulty : fixture.team_a_difficulty
				} else {
					if (difficultyType === 'fpl') {
						difficulty = isHome ? fixture.team_h_difficulty : fixture.team_a_difficulty
					} else {
						// Try to get our calculated FDR
						const teamSpecificKey = `${team.id}-${fixture.id}-${gameweek}`
						const calculatedDifficulty = studioFDRCache.get(teamSpecificKey)

						if (calculatedDifficulty !== undefined) {
							difficulty = calculatedDifficulty
						} else {
							// Fallback to FPL rating
							difficulty = isHome
								? fixture.team_h_difficulty
								: fixture.team_a_difficulty
							console.warn(
								`No calculated difficulty for ${teamSpecificKey}, using FPL rating`,
							)
						}
					}
				}

				return {
					label,
					difficulty: Number(difficulty.toFixed(2)),
					opponentName: opponent?.name ?? 'Unknown',
					kickoffTime: fixture.kickoff_time,
				}
			})

			row.push(fixturesForWeek)
		}

		return row
	})

	// Calculate improved attractiveness scores
	const scores: number[] = teams.map((_, teamIndex) => {
		const teamFixtures = fixtureMatrix[teamIndex]
		const attractivenessResult = calculateAttractivenessScore(teamFixtures)
		return attractivenessResult.totalScore
	})

	return { teamNames, fixtureMatrix, scores }
}
