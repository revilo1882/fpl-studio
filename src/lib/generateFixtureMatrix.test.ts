import { describe, it, expect } from 'vitest'

import type { Team, Fixtures } from '@/types/fpl'

import { generateFixtureMatrix, type IGenerateFixtureMatrix } from './generateFixtureMatrix'

const teams: Team[] = [
	{
		id: 1,
		name: 'Arsenal',
		short_name: 'ARS',
		strength: 4,
	},
	{
		id: 2,
		name: 'Man City',
		short_name: 'MCI',
		strength: 5,
	},
]

const fixtures: Fixtures = [
	{
		id: 1001,
		event: 1,
		team_a: 2,
		team_a_difficulty: 2,
		team_a_score: null,
		team_h: 1,
		team_h_difficulty: 3,
		team_h_score: null,
		kickoff_time: '2024-08-12T15:00:00Z',
		started: false,
		finished: false,
	},
	{
		id: 1002,
		event: 2,
		team_a: 1,
		team_a_difficulty: 1,
		team_a_score: null,
		team_h: 2,
		team_h_difficulty: 4,
		team_h_score: null,
		kickoff_time: '2024-08-19T15:00:00Z',
		started: false,
		finished: false,
	},
]

describe('generateFixtureMatrix', () => {
	it('returns an object with fixtureMatrix, teamNames, and averages', () => {
		const result: IGenerateFixtureMatrix = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result).toBeDefined()
		expect(result).toHaveProperty('fixtureMatrix')
		expect(result).toHaveProperty('teamNames')
		expect(result).toHaveProperty('averages')
	})

	it('returns the correct number of teams in fixtureMatrix and teamNames', () => {
		const result = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result.fixtureMatrix.length).toBe(teams.length)
		expect(result.teamNames.length).toBe(teams.length)
	})

	it('each team row in fixtureMatrix has correct number of gameweeks', () => {
		const gameweeks = 2
		const result = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: gameweeks,
		})
		for (const row of result.fixtureMatrix) {
			expect(row.length).toBe(gameweeks)
		}
	})

	it('averages array length matches number of teams', () => {
		const result = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result.averages.length).toBe(teams.length)
		for (const avg of result.averages) {
			expect(typeof avg).toBe('number')
		}
	})

	it('handles empty fixtures gracefully', () => {
		const result = generateFixtureMatrix({
			fixtures: [],
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result.fixtureMatrix.length).toBe(teams.length)
		for (const row of result.fixtureMatrix) {
			for (const cell of row) {
				expect(cell).toEqual([{ label: '-', difficulty: 0 }])
			}
		}
		// Averages should all be 0 or whatever your function returns for empty
	})

	it('handles empty teams gracefully', () => {
		const result = generateFixtureMatrix({
			fixtures,
			teams: [],
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result.fixtureMatrix.length).toBe(0)
		expect(result.teamNames.length).toBe(0)
		expect(result.averages.length).toBe(0)
	})

	it('returns correct output for single gameweek', () => {
		const result = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 2,
			numberOfGameweeks: 1,
		})
		expect(result.fixtureMatrix.length).toBe(teams.length)
		for (const row of result.fixtureMatrix) {
			expect(row.length).toBe(1)
		}
	})

	it('handles double gameweeks (multiple fixtures for one team in one GW)', () => {
		const dgFixtures: Fixtures = [
			...fixtures,
			{
				id: 1003,
				event: 1,
				team_a: 1,
				team_a_difficulty: 3,
				team_a_score: null,
				team_h: 2,
				team_h_difficulty: 4,
				team_h_score: null,
				kickoff_time: '2024-08-13T15:00:00Z',
				started: false,
				finished: false,
			},
		]
		const result = generateFixtureMatrix({
			fixtures: dgFixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result.fixtureMatrix.length).toBe(teams.length)
		expect(Array.isArray(result.fixtureMatrix[0][0])).toBe(true)
	})

	it('matches the expected output snapshot', () => {
		const result = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})
		expect(result).toMatchSnapshot()
	})

	it('handles double gameweeks (DGW) and stores multiple fixtures as an array in the matrix cell', () => {
		const dgFixtures: Fixtures = [
			{
				id: 1001,
				event: 1,
				team_a: 2,
				team_a_difficulty: 2,
				team_a_score: null,
				team_h: 1,
				team_h_difficulty: 3,
				team_h_score: null,
				kickoff_time: '2024-08-12T15:00:00Z',
				started: false,
				finished: false,
			},
			{
				id: 1002,
				event: 2,
				team_a: 1,
				team_a_difficulty: 1,
				team_a_score: null,
				team_h: 2,
				team_h_difficulty: 4,
				team_h_score: null,
				kickoff_time: '2024-08-19T15:00:00Z',
				started: false,
				finished: false,
			},
			{
				id: 1003,
				event: 1, // Second fixture for Arsenal in GW1 (DGW)
				team_a: 1,
				team_a_difficulty: 2,
				team_a_score: null,
				team_h: 2,
				team_h_difficulty: 4,
				team_h_score: null,
				kickoff_time: '2024-08-13T19:00:00Z',
				started: false,
				finished: false,
			},
		]

		const result = generateFixtureMatrix({
			fixtures: dgFixtures,
			teams: [
				{
					id: 1,
					name: 'Arsenal',
					short_name: 'ARS',
					strength: 4,
				},
				{
					id: 2,
					name: 'Man City',
					short_name: 'MCI',
					strength: 5,
				},
			],
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})

		expect(result.fixtureMatrix.length).toBe(2)
		expect(result.teamNames).toEqual(['Arsenal', 'Man City'])
		expect(Array.isArray(result.fixtureMatrix[0][0])).toBe(true) // Arsenal's GW1
		expect(result.fixtureMatrix[0][0].length).toBe(2) // Two fixtures for Arsenal in GW1
		expect(Array.isArray(result.fixtureMatrix[1][0])).toBe(true) // Man City's GW1 fixtures (should be one)
		expect(result.fixtureMatrix[1][0].length).toBe(2)

		expect(result.fixtureMatrix[0][0][0]).toHaveProperty('difficulty')
		expect(result.averages.length).toBe(2)
		expect(typeof result.averages[0]).toBe('number')
		expect(typeof result.averages[1]).toBe('number')
	})

	it('handles triple gameweeks (TGW) for a team', () => {
		const tgwFixtures: Fixtures = [
			// Arsenal with three fixtures in GW1
			{
				id: 2001,
				event: 1,
				team_a: 2,
				team_a_difficulty: 2,
				team_a_score: null,
				team_h: 1,
				team_h_difficulty: 3,
				team_h_score: null,
				kickoff_time: '2024-08-12T15:00:00Z',
				started: false,
				finished: false,
			},
			{
				id: 2002,
				event: 1,
				team_a: 3,
				team_a_difficulty: 4,
				team_a_score: null,
				team_h: 1,
				team_h_difficulty: 3,
				team_h_score: null,
				kickoff_time: '2024-08-13T19:00:00Z',
				started: false,
				finished: false,
			},
			{
				id: 2003,
				event: 1,
				team_a: 1,
				team_a_difficulty: 3,
				team_a_score: null,
				team_h: 4,
				team_h_difficulty: 4,
				team_h_score: null,
				kickoff_time: '2024-08-14T19:00:00Z',
				started: false,
				finished: false,
			},
		]

		const multiTeams: Team[] = [
			{ id: 1, name: 'Arsenal', short_name: 'ARS', strength: 4 },
			{ id: 2, name: 'Man City', short_name: 'MCI', strength: 5 },
			{ id: 3, name: 'Chelsea', short_name: 'CHE', strength: 4 },
			{ id: 4, name: 'Spurs', short_name: 'TOT', strength: 4 },
		]

		const result = generateFixtureMatrix({
			fixtures: tgwFixtures,
			teams: multiTeams,
			firstGameweek: 1,
			numberOfGameweeks: 1,
		})

		// Arsenal should have 3 fixtures in GW1
		expect(result.fixtureMatrix[0][0].length).toBe(3)

		// Other teams should have at least one, or possibly empty depending on fixture data
		expect(Array.isArray(result.fixtureMatrix[1][0])).toBe(true)
	})

	it('calculates correct difficulty values in matrix', () => {
		const result = generateFixtureMatrix({
			fixtures,
			teams,
			firstGameweek: 1,
			numberOfGameweeks: 2,
		})

		const arsenalGW1Fixtures = result.fixtureMatrix[0][0]
		expect(Array.isArray(arsenalGW1Fixtures)).toBe(true)

		expect(arsenalGW1Fixtures.length).toBe(1)
		expect(arsenalGW1Fixtures[0].difficulty).toBe(3)

		const manCityGW1Fixtures = result.fixtureMatrix[1][0]
		expect(Array.isArray(manCityGW1Fixtures)).toBe(true)
		expect(manCityGW1Fixtures.length).toBe(1)
		expect(manCityGW1Fixtures[0].difficulty).toBe(2)
	})

	it('handles a blank gameweek for a team', () => {
		const blankGWFixtures: Fixtures = [
			{
				id: 3001,
				event: 1,
				team_a: 2,
				team_a_difficulty: 2,
				team_a_score: null,
				team_h: 1,
				team_h_difficulty: 3,
				team_h_score: null,
				kickoff_time: '2024-08-12T15:00:00Z',
				started: false,
				finished: false,
			},
			// No event 2 fixtures for any team
		]

		const multiTeams: Team[] = [
			{ id: 1, name: 'Arsenal', short_name: 'ARS', strength: 4 },
			{ id: 2, name: 'Man City', short_name: 'MCI', strength: 5 },
			{ id: 3, name: 'Chelsea', short_name: 'CHE', strength: 4 },
		]

		const result = generateFixtureMatrix({
			fixtures: blankGWFixtures,
			teams: multiTeams,
			firstGameweek: 2, // GW2 has no fixtures!
			numberOfGameweeks: 1,
		})

		// All teams should have the default "blank" cell in GW2
		multiTeams.forEach((_, idx) => {
			expect(result.fixtureMatrix[idx][0]).toEqual([{ label: '-', difficulty: 0 }])
		})
	})

	it('handles fixtures with finished games and scores', () => {
		const scoredFixtures: Fixtures = [
			{
				id: 4001,
				event: 1,
				team_a: 2,
				team_a_difficulty: 2,
				team_a_score: 2,
				team_h: 1,
				team_h_difficulty: 3,
				team_h_score: 1,
				kickoff_time: '2024-08-12T15:00:00Z',
				started: true,
				finished: true,
			},
		]

		const twoTeams: Team[] = [
			{ id: 1, name: 'Arsenal', short_name: 'ARS', strength: 4 },
			{ id: 2, name: 'Man City', short_name: 'MCI', strength: 5 },
		]

		const result = generateFixtureMatrix({
			fixtures: scoredFixtures,
			teams: twoTeams,
			firstGameweek: 1,
			numberOfGameweeks: 1,
		})

		expect(result.fixtureMatrix[0][0][0]).toMatchObject({
			label: 'MCI (H)',
			difficulty: 3,
		})
		expect(result.fixtureMatrix[1][0][0]).toEqual({
			label: 'ARS (A)',
			difficulty: 2,
		})
	})
})
