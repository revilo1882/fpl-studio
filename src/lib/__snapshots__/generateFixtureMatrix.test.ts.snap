// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateFixtureMatrix > matches the expected output snapshot 1`] = `
{
  "averages": [
    2,
    3,
  ],
  "fixtureMatrix": [
    [
      [
        {
          "difficulty": 3,
          "label": "MCI (H)",
        },
      ],
      [
        {
          "difficulty": 1,
          "label": "MCI (A)",
        },
      ],
    ],
    [
      [
        {
          "difficulty": 2,
          "label": "ARS (A)",
        },
      ],
      [
        {
          "difficulty": 4,
          "label": "ARS (H)",
        },
      ],
    ],
  ],
  "teamNames": [
    "Arsenal",
    "Man City",
  ],
}
`;
